File: ruby/run-length-encoding/run_length_encoding.rb
module BookKeeping
    VERSION = 3
end
  
#more then inspired by other solutions
module RunLengthEncoding
        
  def self.encode str
    str.chars.chunk(&:itself).map { |x,arr|  
    (arr.size > 1 ? arr.size.to_s : '') + x }.join
  end
      
  def self.decode str
    str.scan(/(\d*)(\D)/)
       .reduce(String.new) { |decoded, (num, char) |  
         decoded << (num.empty? ? char : char * num.to_i) }
  end

end

File: javascript/run-length-encoding/run-length-encoding.js
class RunLengthEncoding {

    encode(str) {
        if (str == '') return '';
        let encoded = str.match(/(.)\1*/g).map(x => x.length + x[0]).join('');
        return encoded.split(/(\d*)/g).filter(x => x != 1).join('');
      }
  
    decode(str) {
        return str
            .split(/(\d*)/g)
            .filter(x => x !== '')
            .reduce(
                (arr, x) => {
                    if (Number(x)) {
                        arr.num = Number(x);
                        return arr;
                    } else {
                        arr.str = arr.str + x.repeat(arr.num);
                        arr.num = 1;
                        return arr;
                    }
                },
                { str: '', num: 1 }
            ).str;
    }
}

var rle = new RunLengthEncoding();

module.exports = rle;